search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,433 @@
+                                     (GSList **) &models,
+                                     (GSList **) &shards,
+                                     &error))
+    {
+      g_dbus_method_invocation_take_error (state->invocation,
+                                           g_steal_pointer (&error));
+      return;
+    }
+
+  shards_strv = strv_from_shard_list ((GSList *) shards);
+
+  for (GSList *l = (GSList *) models; l; l = l->next)
+    {
+      EkncContentObjectModel *model = l->data;
+
+      g_variant_builder_open (&builder, G_VARIANT_TYPE("a{sv}"));
+
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "title",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "content-type",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "thumbnail-uri",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "ekn-id",
+                                                      G_VARIANT_TYPE ("s"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "child-tags",
+                                                      G_VARIANT_TYPE ("as"));
+      maybe_add_key_value_pair_from_model_to_variant (model,
+                                                      &builder,
+                                                      "tags",
+                                                      G_VARIANT_TYPE ("as"));
+
+      g_variant_builder_close (&builder);
+    }
+
+  eks_article_metadata_complete_query (state->provider->skeleton,
+                                       state->invocation,
+                                       (const gchar * const *) shards_strv,
+                                       g_variant_builder_end (&builder));
+}
+
+static GValue *
+append_construction_prop (const gchar *key,
+                          GType        type,
+                          GArray      *values_array,
+                          GPtrArray   *props_array,
+                          guint       *nprops)
+{
+  guint index = values_array->len;
+  g_array_set_size (values_array, values_array->len + 1);
+  g_value_init (&g_array_index (values_array, GValue, index),
```

Comment:
You can probably factor out the gvalue pointer since it's used in two places

Reply:
True that